bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/handler/CaddxBaseThingHandler.java

```diff
@@ -0,0 +1,309 @@
+import org.eclipse.smarthome.core.thing.Bridge;
+import org.eclipse.smarthome.core.thing.Channel;
+import org.eclipse.smarthome.core.thing.ChannelUID;
+import org.eclipse.smarthome.core.thing.Thing;
+import org.eclipse.smarthome.core.thing.ThingStatus;
+import org.eclipse.smarthome.core.thing.ThingStatusDetail;
+import org.eclipse.smarthome.core.thing.ThingStatusInfo;
+import org.eclipse.smarthome.core.thing.binding.BaseThingHandler;
+import org.eclipse.smarthome.core.thing.binding.ThingHandler;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.caddx.internal.CaddxEvent;
+import org.openhab.binding.caddx.internal.config.CaddxKeypadConfiguration;
+import org.openhab.binding.caddx.internal.config.CaddxPartitionConfiguration;
+import org.openhab.binding.caddx.internal.config.CaddxZoneConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Abstract class for a Caddx Thing Handler.
+ *
+ * @author Georgios Moutsos - Initial contribution
+ */
+@NonNullByDefault
+public abstract class CaddxBaseThingHandler extends BaseThingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(CaddxBaseThingHandler.class);
+
+    /** Bridge Handler for the Thing. */
+    private @Nullable CaddxBridgeHandler caddxBridgeHandler = null;
+
+    /** Caddx Alarm Thing type. */
+    private CaddxThingType caddxThingType;// = null;
+
+    /** Caddx Properties. */
+    private boolean thingHandlerInitialized = false;
+
+    /** Partition Number. */
+    private int partitionNumber;
+
+    /** Zone Number. */
+    private int zoneNumber;
+
+    /** Keypad Address. */
+    private int keypadAddress;
+
+    /**
+     * Constructor.
+     *
+     * @param thing
+     */
+    /*
+     * public CaddxBaseThingHandler(Thing thing) {
+     * super(thing);
+     * }
+     */
+    public CaddxBaseThingHandler(Thing thing, CaddxThingType caddxThingType) {
+        super(thing);
+        this.caddxThingType = caddxThingType;
+        getCaddxBridgeHandler();
+    }
+
+    @Override
+    public void initialize() {
+        logger.trace("initialize(): {}; Thing ID: {}.", caddxThingType, this.getThing().getUID());
+
+        getConfiguration(caddxThingType);
+
+        // set the Thing offline for now
+        updateStatus(ThingStatus.OFFLINE);
+    }
+
+    @Override
+    public void dispose() {
+        logger.trace("Thing {} disposed.", getThing().getUID());
+
+        this.setThingHandlerInitialized(false);
+
+        super.dispose();
+    }
+
+    /**
+     * Method to Initialize Thing Handler.
+     */
+    public void initializeThingHandler() {
+        logger.trace("initializeThingHandler()");
+
+        if (getCaddxBridgeHandler() != null) {
+            this.setThingHandlerInitialized(true);
+        }
+    }
+
+    /**
+     * Get the Bridge Handler for the Caddx system.
+     *
+     * @return CaddxBridgeHandler
+     */
+    public @Nullable synchronized CaddxBridgeHandler getCaddxBridgeHandler() {
```

Comment:
why is this synchronized?

Reply:
To not enter inside the if twice. (I do not know if this is possible)
As far as i can tell no harm will be done even if it enters twice. 
But i decided to think defensive.
Should i remove it?

Reply:
It is safe to remove it.

Reply:
Done